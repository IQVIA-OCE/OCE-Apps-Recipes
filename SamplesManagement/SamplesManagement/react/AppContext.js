import React, { Component } from 'react';
import { sfNetAPI } from './bridge/sf/sfnetapi';
import { environment } from './bridge/EnvironmentData/EnvironmentData.native';
import { View } from 'react-native';
import { ActivityIndicator, Colors } from 'apollo-react-native';

const defaultContextValue = {
  isLoading: true,
  userId: null,
  profileId: null,
  orgId: null,
  error: null,
};

export const AppContext = React.createContext(defaultContextValue);

export default class extends Component {
  state = defaultContextValue;

  fetchUser = () => {
    const userId = environment.userID();
    return new Promise((resolve, reject) => {
      sfNetAPI.query(
        `SELECT Id, Name, OCE__ProfileId__c FROM User WHERE Id = '${userId}'`,
        data => {
          // console.log(data);
          resolve(data.records);
        },
        e => {
          // console.log(e);
          reject(e);
        }
      );
    });
  };

  fetchOrg = () => {
    return new Promise((resolve, reject) => {
      sfNetAPI.query(
        `SELECT Id FROM Organization`,
        data => {
          // console.log(data);
          resolve(data.records);
        },
        e => {
          // console.log(e);
          reject(e);
        }
      );
    });
  };

  getInitialData = async () => {
    this.setState({ isLoading: true });
    try {
      const [user] = await this.fetchUser();
      const [org] = await this.fetchOrg();

      this.setState({
        profileId: user.OCE__ProfileId__c,
        username: user.Name,
        orgId: org.Id,
        userId: environment.userID(),
        isLoading: false,
        error: null,
      });
    } catch (error) {
      console.warn(error);
      this.setState({ isLoading: false, error });
    }
  };

  componentDidMount() {
    this.getInitialData();
  }

  render() {
    return (
      <AppContext.Provider value={this.state}>
        {this.state.isLoading ? (
          <View style={{ padding: 20 }}>
            <ActivityIndicator
              animating={true}
              color={Colors.blue}
              style={{ paddingVertical: 10 }}
            />
          </View>
        ) : (
          this.props.children
        )}
      </AppContext.Provider>
    );
  }
}
