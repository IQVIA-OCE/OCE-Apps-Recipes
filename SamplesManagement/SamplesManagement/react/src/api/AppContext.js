import api from '../utils/api';

export const fetchUser = userId =>
  api.query(
    `SELECT Id, Name, OCE__ProfileId__c FROM User WHERE Id = '${userId}'`
  );