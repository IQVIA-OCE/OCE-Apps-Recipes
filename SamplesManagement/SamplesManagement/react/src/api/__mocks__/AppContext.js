export const fetchUser = jest.fn().mockResolvedValue([
  [
    {
      Id: '0050k000004CineAAC',
      Name: 'SALESREP',
      OCE__ProfileId__c: '00e0k000000O1zxAAC',
    },
  ],
  {
    totalSize: 1,
    done: true,
  },
]);
