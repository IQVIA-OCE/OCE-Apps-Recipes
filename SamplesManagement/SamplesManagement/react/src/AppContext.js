import React, { Component } from 'react';
import { sfNetAPI, environment } from '@oce-apps/oce-apps-bridges';
import { View } from 'react-native';
import { ActivityIndicator, withTheme } from '@oce-apps/apollo-react-native';
import { fetchUser } from './api/AppContext';
import { NAMESPACE } from './constants/constants';

const defaultContextValue = {
  isLoading: true,
  userId: null,
  profileId: null,
  orgId: null,
  error: null,
};

export const AppContext = React.createContext(defaultContextValue);

class AppContextProvider extends Component {
  state = defaultContextValue;

  getInitialData = async () => {
    try {
      const userId = environment.userId();
      const [[user]] = await fetchUser(userId);

      this.setState({
        profileId: user[`${NAMESPACE}ProfileId__c`],
        username: user.Name,
        isLoading: false,
        error: null,
        userId
      });
    } catch (error) {
      this.setState({ isLoading: false, error });
    }
  };

  componentDidMount() {
    this.getInitialData();
  }

  render() {
    const { theme } = this.props;

    return (
      <AppContext.Provider value={this.state}>
        {this.state.isLoading ? (
          <View style={{ flex: 1, justifyContent: 'center', alignItems: 'center', backgroundColor: theme.colors.background }}>
            <ActivityIndicator
              animating={true}
              color={theme.colors.primary}
              style={{ paddingVertical: 10 }}
            />
          </View>
        ) : (
          this.props.children
        )}
      </AppContext.Provider>
    );
  }
}

export default withTheme(AppContextProvider);
