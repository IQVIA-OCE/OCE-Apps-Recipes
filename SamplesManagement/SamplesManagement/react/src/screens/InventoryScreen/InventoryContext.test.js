import React, { useContext } from 'react';
import { act, render } from '@testing-library/react-native';
import {
  fetchActiveLotsProducts,
  fetchInventoryById,
} from '../../api/Inventories';
import InventoryContextProvider, { InventoryContext } from './InventoryContext';
import { Text } from 'react-native';
import { AppContext } from '../../AppContext';
import { customSettings } from '@oce-apps/oce-apps-bridges';
import { useRoute } from '@react-navigation/native';

jest.mock('@react-navigation/native');

describe('InventoryContext', () => {
  beforeAll(() => {
    customSettings.getCustomSetting.mockResolvedValue([{ Id: 1 }]);
  });
  beforeEach(() => {
    fetchInventoryById.mockClear();
  });

  it('should render properly with error', () => {
    useRoute.mockReturnValue({
      params: {
        id: '1'
      }
    });
    fetchActiveLotsProducts.mockRejectedValueOnce('error');

    const InnerComponent = () => {
      const context = useContext(InventoryContext);

      if (context.error) return <Text>ContextError</Text>

      return null;
    };

    const Component = () => (
      <AppContext.Provider value={{}}>
        <InventoryContextProvider>
          <InnerComponent />
        </InventoryContextProvider>
      </AppContext.Provider>
    );

    const { getByText } = render(<Component />);
    act(() => jest.runAllTimers());

    expect(getByText(/ContextError/)).toBeTruthy()
  });

  it('should render with create type', () => {
    useRoute.mockReturnValue({
      params: {}
    });

    const InnerComponent = () => {
      const context = useContext(InventoryContext);

      return <Text>{JSON.stringify({ context })}</Text>;
    };

    const Component = () => (
      <AppContext.Provider value={{}}>
        <InventoryContextProvider navigation={{ state: { params: {} } }}>
          <InnerComponent />
        </InventoryContextProvider>
      </AppContext.Provider>
    );

    render(<Component />);
    act(() => jest.runAllTimers());

    expect(fetchInventoryById).not.toHaveBeenCalled()
  });

  it('should render with edit type', () => {
    useRoute.mockReturnValue({
      params: {
        id: '1'
      }
    })

    const InnerComponent = () => {
      const context = useContext(InventoryContext);

      return <Text>{JSON.stringify({ context })}</Text>;
    };
    const Component = () => (
      <AppContext.Provider value={{}}>
        <InventoryContextProvider navigation={{ state: { params: {id: '1'} } }}>
          <InnerComponent />
        </InventoryContextProvider>
      </AppContext.Provider>
    );

    render(<Component />);
    act(() => jest.runAllTimers());

    expect(fetchInventoryById).toHaveBeenCalledTimes(1);
  });
});
