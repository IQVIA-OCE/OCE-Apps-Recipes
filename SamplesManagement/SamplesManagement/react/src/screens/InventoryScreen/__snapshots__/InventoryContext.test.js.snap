// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InventoryContext should render properly with error 1`] = `
<View>
  <Text>
    Inventory
  </Text>
</View>
`;

exports[`InventoryContext should render with create type 1`] = `
<Text>
  {"context":{"editingType":2,"isLoading":false,"preselectedProducts":[],"products":[{"Id":1}],"error":null,"config":{"Id":1},"transactionDetails":[{"Id":1}],"formInitialValues":{"id":"","datetime":"2020-11-23T18:43:24.926Z","products":[{"Id":1}],"status":"In Progress","buttonPressed":""},"actions":{}}}
</Text>
`;

exports[`InventoryContext should render with edit type 1`] = `
<Text>
  {"context":{"editingType":0,"isLoading":false,"preselectedProducts":[],"products":[{"Id":1}],"error":null,"config":{"Id":1},"transactionDetails":[{"Id":1}],"formInitialValues":{"Id":1,"products":[{"Id":1}],"buttonPressed":""},"actions":{}}}
</Text>
`;
